File: samples/abro/ABRO.java
package samples.abro;

import java.util.List;
import samples.abro.abstract_base_classes.State;
import samples.abro.abstract_base_classes.Region;

public class ABRO extends State {
    public boolean A, B, R, O;

    public ABRO() {
        super(false);
        this.regions = List.of(new Root());
        reset();
    }

    @Override
    public void localReset() {
        A = false;
        B = false;
        R = false;
        O = false;
    }

    class Root extends Region {

        private final State ABO;

        public Root() {
            this.ABO = new ABO();
            this.initialState = ABO;
            this.states = List.of(ABO);
        }

        @Override
        public boolean didStrongAborts() {
            if (activeState.equals(ABO)) {
                if (delayedEnabled && R) { // if R abort to ABO
                    transitionTo(ABO);
                    return true;
                }
            }
            return false;
        }

        class ABO extends State {

            public ABO() {
                super(false);
                this.regions = List.of(new Wait());
            }

            @Override
            public void onEntry() {
                O = false; // entry do O = false
            }

            class Wait extends Region {

                private final State WaitAB, Done;

                public Wait() {
                    this.WaitAB = new WaitAB();
                    this.Done = new State(false);

                    this.initialState = WaitAB;
                    this.states = List.of(WaitAB, Done);
                }

                @Override
                public boolean didWeakAborts() {
                    if (activeState.equals(WaitAB)) {
                        if (activeState.isTerminated()) { // do O = true join to Done
                            transitionTo(Done, () -> { O = true; });
                            return true;
                        }
                    }
                    return false;
                }

                class WaitAB extends State {

                    public WaitAB() {
                        super(false);
                        this.regions = List.of(new HandleA(), new HandleB());
                    }

                    class HandleA extends Region {

                        private final State WaitA, DoneA;

                        public HandleA() {
                            this.WaitA = new State(false);
                            this.DoneA = new State(true);

                            this.initialState = WaitA;
                            this.states = List.of(WaitA, DoneA);
                        }

                        @Override
                        public boolean didWeakAborts() {
                            if (activeState.equals(WaitA)) {
                                if (this.delayedEnabled && A) { // if A go to DoneA
                                    transitionTo(DoneA);
                                    return true;
                                }
                            }
                            return false;
                        }
                    }

                    class HandleB extends Region {

                        private final State WaitB, DoneB;

                        public HandleB() {
                            this.WaitB = new State(false);
                            this.DoneB = new State(true);

                            this.initialState = WaitB;
                            this.states = List.of(WaitB, DoneB);
                        }

                        @Override
                        public boolean didWeakAborts() {
                            if (activeState.equals(WaitB)) {
                                if (this.delayedEnabled && B) { // if B go to DoneB
                                    transitionTo(DoneB);
                                    return true;
                                }
                            }
                            return false;
                        }
                    }
                }
            }
        }
    }
}

File: samples/abro_ts/src/abro.ts
import {
    State,
    Region,
    SimpleState,
    RegionImpl,
    StateImpl,
} from "./base_classes/base";

const ABRO = (_isFinal: boolean) => {
    let A = false;
    let B = false;
    let R = false;
    let O = false;

    const localReset = () => {
        A = false;
        B = false;
        R = false;
        O = false;
    };

    const context = {
        get A() { return A; },
        set A(value: boolean) { A = value; },
        get B() { return B; },
        set B(value: boolean) { B = value; },
        get R() { return R; },
        set R(value: boolean) { R = value; },
        get O() { return O; },
        set O(value: boolean) { O = value; },
    };

    const Root: Region = (() => {
        let ABO: State = ((_isFinal: boolean) => {
            const Wait: Region = (() => {
                const WaitAB: State = ((_isFinal: boolean) => {
                    const HandleA: Region = (() => {
                        const WaitA: State = new SimpleState(false);
                        const DoneA: State = new SimpleState(true);

                        const states: State[] = [WaitA, DoneA];
                        const initialState: State = WaitA;

                        const handleNonPreemptiveTransitions = function (
                            this: RegionImpl,
                        ): boolean {
                            if (this.activeState === WaitA) {
                                if (A) {
                                    this.transitionTo(DoneA);
                                    return true;
                                }
                            }
                            return false;
                        };

                        return new RegionImpl(states, initialState, {
                            handleNonPreemptiveTransitions,
                        });
                    })();

                    const HandleB: Region = (() => {
                        const WaitB: State = new SimpleState(false);
                        const DoneB: State = new SimpleState(true);

                        const states: State[] = [WaitB, DoneB];
                        const initialState: State = WaitB;

                        const handleNonPreemptiveTransitions = function (
                            this: RegionImpl,
                        ): boolean {
                            if (this.activeState === WaitB) {
                                if (B) {
                                    this.transitionTo(DoneB);
                                    return true;
                                }
                            }
                            return false;
                        };

                        return new RegionImpl(states, initialState, {
                            handleNonPreemptiveTransitions,
                        });
                    })();

                    const regions: Region[] = [HandleA, HandleB];

                    return new StateImpl(_isFinal, regions);
                })(false);

                const Done: State = new SimpleState(true);

                const states: State[] = [WaitAB, Done];
                const initialState: State = WaitAB;

                const handleNonPreemptiveTransitions = function (
                    this: RegionImpl,
                ): boolean {
                    if (this.activeState === WaitAB) {
                        if (this.activeState.isTerminated()) {
                            this.transitionTo(Done, () => {
                                O = true;
                            });
                            return true;
                        }
                    }
                    return false;
                };

                return new RegionImpl(states, initialState, {
                    handleNonPreemptiveTransitions,
                });
            })();

            const regions: Region[] = [Wait];

            const onEntry = function onEntry(): void {
                O = false;
            };

            return new StateImpl(_isFinal, regions, { onEntry });
        })(false);

        const states: State[] = [ABO];
        const initialState: State = ABO;

        const handlePreemptiveTransitions = function (this: RegionImpl): boolean {
            if (this.activeState === ABO) {
                if (R) {
                    this.transitionTo(ABO);
                    return true;
                }
            }
            return false;
        };

        return new RegionImpl(states, initialState, {
            handlePreemptiveTransitions,
        });
    })();

    const regions: Region[] = [Root];

    const model: State = new StateImpl(_isFinal, regions, { localReset });

    return {
        model,
        context,
    };
};

export { ABRO };
